panther_analysis_tool/test_case.py

```diff
@@ -0,0 +1,75 @@
+from jsonpath_ng import Fields, parse
+
+
+# pylint: disable=too-few-public-methods
+class DataModel:
+
+    def __init__(self, data_model_id: str,
+                 source_mappings: List[Dict[str, str]], mapping_module: Any):
+        self.data_model_id: str = data_model_id
+        self.paths: Dict[str, Fields] = dict()
+        self.methods: Dict[str, Callable] = dict()
+        self.module = mapping_module
+        self.extract_mappings(source_mappings)
+
+    def extract_mappings(self, source_mappings: List[Dict[str, str]]) -> None:
+        for mapping in source_mappings:
+            # every mapping should have a name and either a method or path
+            if 'Method' in mapping:
+                self.methods[mapping['Name']] = getattr(self.module,
+                                                        mapping['Method'])
+            else:
+                self.paths[mapping['Name']] = parse(mapping['Path'])
+
+
+class TestCase():
+
+    def __init__(self, data: Dict[str, Any],
+                 data_models: List[DataModel]) -> None:
+        """
+        Args:
+            data (Dict[str, Any]): An AWS Resource representation or Log event to test the policy
+            or rule against respectively.
+        """
+        self._data = data
+        self.data_models = data_models
+
+    def __getitem__(self, arg: str) -> Any:
+        return self._data.get(arg, None)
+
+    def __iter__(self) -> Iterator:
+        return iter(self._data)
+
+    def get(self, arg: str, default: Any = None) -> Any:
+        return self._data.get(arg, default)
+
+    def udm(self, key: str) -> Any:
+        """Converts standard data model field to logtype field"""
+        # access values via normal logType fields
+        if key in self._data.keys():
+            return self.get(key)
+        if self.data_models:
+            # access values via standardized fields
+            # check each data model that could apply
+            for data_model in self.data_models:
+                if key in data_model.paths.keys():
+                    # we are dealing with a jsonpath
+                    json_path = data_model.paths.get(key)
+                    if json_path:
+                        matches = json_path.find(self._data)
+                        if len(matches) == 1:
+                            return matches[0].value
+                        if len(matches) > 1:
+                            raise Exception(
+                                'JSONPath [{}] in DataModel [{}], matched multiple fields.'
+                                .format(json_path, data_model.data_model_id))
+                if key in data_model.methods.keys():
+                    # we are dealing with method
+                    method = data_model.methods.get(key)
+                    if callable(method):
+                        result = method(self._data)
+                        if result:
+                            return result
```

Comment:
Yeah this is a sticky issue. Technically `p_log_type` is a required field in all logs, we just typically skip it in our test cases. If we said all unit tests needed to specify `p_log_type` if a data model is configured for them, that could work (we'd just refer to that field to determine which data model to use).

Kind of makes data models not super backwards compatible with existing rules though. But existing rules won't be using data models anyways, so maybe it's not a big deal?

The more I think about it, the more I think that we should just force `p_log_type` in the `Log` for any rules that use a data model.

Reply:
I like this idea.  However, I'm unsure how to check it until `udm` is being called (when the test is running), which means it would be added to the `failed_tests` rather than `invalid_spec` - I assume we would want to add it to the `invalid_spec` instead? I can look at how its running and re-work it if that's what we would prefer.

Reply:
The more I think about it - the more I think `.udm()` should only ever return values for the unified data model and any time users want access to the event fields, they have to do `.get()` (or [] access).  I was trying to make it easier on the user so that they could use `.udm()` for either, but it makes dealing with these collisions more difficult.  Thoughts?

Reply:
I think `failed_tests` is fine, it indicates that the tests are not correct. In this case it just means that the test is missing a  field which is arguably an `invalid_spec`, but I don't think we should knock ourselves out of this one unless customers run into  repeated issues with it.